File: apps/ios/MiPlancha/AuthService.swift
//
//  AuthService.swift
//  MiPlancha
//
//  Login methods.
//
//  Created by julio on 21/08/2026.
//

import Observation
import FirebaseAuth
import FirebaseFirestore

/// Service to manage authentication
///
/// MainActor: Indicates that all methods here will run in the main thread. This is necessary at any observable object which touches the UI.
///
@Observable
@MainActor
final class AuthService {
    private(set) var user: User?
    private(set) var usuario: Usuario?
    
    init() {
        user = Auth.auth().currentUser
    }
    
    func initSession(companyCode: String, username: String, password: String) async throws {
        let mail = UsernameEmail.syntheticEmail(companyCode: companyCode, username: username)
        let result = try await Auth.auth().signIn(withEmail: mail, password: password)
        self.user = result.user
        self.usuario = try await fetchUser(uid: result.user.uid)
    }
    
    func logout() throws {
        try Auth.auth().signOut()
        user = nil
        usuario = nil
    }
    
    private func fetchUser(uid: String) async throws -> Usuario {
        let snap = try await Firestore.firestore().collection("usuarios").document(uid).getDocument()
        guard let data = snap.data(),
              let name = data["nombre"] as? String,
              let rolRaw = data["rol"] as? String,
              let rol = Rol(rawValue: rolRaw)
        else { throw AuthError.invalidData }
        
        let companyId = data["empresaId"] as? String
        
        return Usuario(uid: uid, name: name, rol: rol, companyId: companyId)
    }
}

enum AuthError: Error {
    case invalidData
}

File: apps/android/app/src/main/java/com/jugomo/miplancha/AuthService.kt
package com.jugomo.miplancha

import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.tasks.await

class AuthService() {
    private var _mutableUser : MutableStateFlow<Usuario?>
    private var _mutableIsRestoring : MutableStateFlow<Boolean>

    lateinit var mutableUser: StateFlow<Usuario?>
    lateinit var mutableIsRestoring : StateFlow<Boolean>

    init {
        _mutableUser = MutableStateFlow(null)
        _mutableIsRestoring = MutableStateFlow(true)

        mutableUser = _mutableUser.asStateFlow()
        mutableIsRestoring =_mutableIsRestoring.asStateFlow()
    }

    suspend fun startSession(companyId: String, username: String, password: String) {
        val syntheticEmail = UsernameEmail.syntheticEmail(companyId, username)
        val res: AuthResult

        try {
            res = FirebaseAuth.getInstance()
                .signInWithEmailAndPassword(syntheticEmail, password).await()

            val uid = res.user?.uid
            if (uid != null) {
                _mutableUser.value = getUser(uid)
            }
        } catch (e: Exception) {
            throw AuthError()
        }
    }

    suspend fun restoreSessionIfActiveUser() {
        _mutableIsRestoring.value = false

        val currUser =  FirebaseAuth.getInstance().currentUser
        if(currUser != null) {
            try {
                _mutableUser.value  = getUser(currUser.uid)
            } catch (e: Exception) {
                throw AuthError()
            }
        }
    }

     fun closeSession() {
        FirebaseAuth.getInstance().signOut()
        _mutableUser.value = null
    }

    private suspend fun getUser(uid: String): Usuario {
        val userSnapshot = FirebaseFirestore.getInstance()
            .collection("usuarios")
            .document(uid).get().await();

        val user = userSnapshot.toUsuario(
            uid = uid
        )

        return user
    }
}

class AuthError() : Exception("Auth Exception") {
}